src/v/kafka/server/quota_manager.cc

```diff
@@ -350,19 +359,28 @@ void quota_manager::gc() {
       = ssx::spawn_with_gate_then(
           _gate,
           [full_window, _client_quotas = std::ref(_client_quotas)]() {
-              auto now = clock::now();
-              auto expire_age = full_window * 10;
-              auto new_map = client_quotas_map_t{*_client_quotas.get().local()};
-              auto erased = absl::erase_if(
-                new_map,
-                [now,
-                 expire_age](const std::pair<ss::sstring, client_quota>& q) {
-                    return (now - q.second.last_seen) > expire_age;
+              return _client_quotas.get().update_async(
+                [full_window](auto old_map) -> ss::future<client_quotas_map_t> {
+                    // TODO: here can hold up the request path by holding the
+                    // sharded_ptr's mutex, which would block requests with new
+                    // client.id's trying to insert into the map. We could
+                    // consider decoupling the gathering of to-be-removed
+                    // entries from the reset and accept losing some recently
+                    // added client ids.
+                    auto new_map = client_quotas_map_t{};
+                    auto now = clock::now();
+                    auto expire_age = full_window * 10;
+                    for (auto kv : old_map) {
+                        auto last_seen_ms
+                          = co_await kv.second.last_seen_ms->max();
```

Comment:
Agreed. I rewrote this now to avoid cross-shard calls inside the `sharded_ptr::update` by first gathering a list of expired keys and then updating the map by removing them like you said.

The downside with my current implementation is that there is a small chance that we incorrectly remove a client id that we (1) first think should be GC'd because we haven't seen for a while (2) but then before we get to updating the map another shard touches this client id again and bumps its `last_seen_ms`. Then we would incorrectly remove that client id. But that seems acceptable given that it should be extremely rare (the request from the client id has to come in exactly 10s after its previous request perfectly racing with `gc()`) and even if this happens we will just end up under-throttling the client.